xarray/core/datatree.py

```diff
@@ -1261,50 +1261,53 @@ def equals(self, other: DataTree, from_root: bool = True) -> bool:
         ----------
         other : DataTree
             The other tree object to compare to.
-        from_root : bool, optional, default is True
-            Whether or not to first traverse to the root of the two trees before checking for isomorphism.
-            If neither tree has a parent then this has no effect.
 
         See Also
         --------
         Dataset.equals
         DataTree.isomorphic
         DataTree.identical
         """
-        if not self.isomorphic(other, from_root=from_root, strict_names=True):
+        if not self.isomorphic(other, strict_names=True):
             return False
 
         return all(
-            [
-                node.dataset.equals(other_node.dataset)
-                for node, other_node in zip(self.subtree, other.subtree, strict=True)
-            ]
+            node.dataset.equals(other_node.dataset)
+            for node, other_node in zip(self.subtree, other.subtree, strict=True)
         )
 
-    def identical(self, other: DataTree, from_root=True) -> bool:
+    def identical(self, other: DataTree) -> bool:
         """
         Like equals, but will also check all dataset attributes and the attributes on
         all variables and coordinates.
 
-        By default this method will check the whole tree above the given node.
-
         Parameters
         ----------
         other : DataTree
             The other tree object to compare to.
-        from_root : bool, optional, default is True
-            Whether or not to first traverse to the root of the two trees before checking for isomorphism.
-            If neither tree has a parent then this has no effect.
 
         See Also
         --------
         Dataset.identical
         DataTree.isomorphic
         DataTree.equals
         """
-        if not self.isomorphic(other, from_root=from_root, strict_names=True):
+        if not self.isomorphic(other, strict_names=True):
+            return False
+
+        if self.name != other.name:
+            return False
+
+        # Check the root node's dataset twice, one with (below) and once without
+        # (here) inheritence. This ensures that even inheritance matches for
+        # identical DataTree objects, although inherited variables need not be
+        # defined at the same level.
+        self_ds = self._to_dataset_view(rebuild_dims=False, inherit=False)
+        other_ds = other._to_dataset_view(rebuild_dims=False, inherit=False)
+        if not self_ds.identical(other_ds):
```

Comment:
Isn't this doing a bunch of equality checking twice unnecessarily? Every non-inherited local coordinate and every local data variable will be checked for equality twice.

We need to check that the set of inherited coordinates are identical, and we then need to check that the set of all other variables are the same. Couldn't we do that with something roughly like

```python
self_ds_inherited = self._to_dataset_view(rebuild_dims=False, inherit=True)
self_ds_non_inherited = self._to_dataset_view(rebuild_dims=False, inherit=False)
self_inherited_coords = set(self_ds_inherited.coords).difference(self_ds_non_inherited.coords)
self_non_inherited_vars = set(self_ds_inherited.variables).intersection(self_ds_non_inherited.variables)

other_ds_inherited = other._to_dataset_view(rebuild_dims=False, inherit=True)
other_ds_non_inherited = other._to_dataset_view(rebuild_dims=False, inherit=False)
other_inherited_coords = set(other_ds_inherited.coords).difference(other_ds_non_inherited.coords)
other_non_inherited_vars = set(other_ds_inherited.variables).intersection(self_ds_non_inherited.variables)

self_inherited_coords.identical(other_inherited_coords)
self_non_inherited_vars.identical(other_non_inherited_vars)
```

Reply:
Maybe this function could be useful

https://github.com/pydata/xarray/pull/9470#discussion_r1751153681

Reply:
Great point, this is indeed quite inefficient. I added a more efficient solution in my latest commit.